batch_builder.go

```diff
@@ -0,0 +1,228 @@
+package hippo
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"sort"
+)
+
+// BatchBuilder is responsible for building a batch that will serialize within a desired size if possible.
+// It'll fit big and small populators together, but doesn't try to do so optimally.
+type BatchBuilder struct {
+	desiredSize       int
+	buf               *bytes.Buffer
+	serializedUpserts [][]byte
+	batchGUID         string
+	replaceAll        bool
+	ttlMinutes        uint32
+	builtBatchesCount int
+	hasClosedBatch    bool
+	sender            TagBatchPartSender // optional - to help track batch origin
+}
+
+// NewBatchBuilder builds a new BatchBuilder
+func NewBatchBuilder(desiredSize int, replaceAll bool, ttlMinutes uint32) *BatchBuilder {
```

Comment:
nit: should we extend this function to include returning a possible error if the desired size is too small? Based on the implementation of `BuildBatch`, providing a desired size of 40 or less would consistently result in errors since there is no space available for upserts, but we wouldn't actually encounter them until we called `BuildBatch`

Reply:
Yeah, I think that's good. You'd get the error later when building the batch. I also considered returning an error when you add an upsert that's too big.